modules/tinymce/src/core/main/ts/newline/InsertBlock.ts

```diff
@@ -242,139 +224,198 @@ const addBrToBlockIfNeeded = function (dom, block) {
   }
 };
 
-const insert = function (editor: Editor, evt?: EditorEvent<KeyboardEvent>) {
-  let tmpRng, container, offset, parentBlock;
-  let newBlock, fragment, containerBlock, parentBlockName, newBlockName, isAfterLastNodeInContainer;
-  const dom = editor.dom;
-  const schema = editor.schema, nonEmptyElementsMap = schema.getNonEmptyElements();
-  const rng = editor.selection.getRng();
 
-  // Creates a new block element by cloning the current one or creating a new one if the name is specified
-  // This function will also copy any text formatting from the parent block and add it to the new one
-  const createNewBlock = function (name?) {
-    let node = container, block, clonedNode, caretNode;
-    const textInlineElements = schema.getTextInlineElements();
-
-    if (name || parentBlockName === 'TABLE' || parentBlockName === 'HR') {
-      block = dom.create(name || newBlockName);
-    } else {
-      block = parentBlock.cloneNode(false);
-    }
+// Check that we aren't accidentally about to extract the entire parentNode.
+const handleEmptyRange = (parentNode: Node, rng: Range) => {
```

Comment:
I'll need to look into this more, as there's a number of things that look odd here. Also not too keen on the mutation of the range here.

- [ ] Reminder for myself to investigate this solution

Reply:
Yeah, the core issue here is that `moveToCaretPosition` is moving the caret before the pre element when pressing enter. That's how it ends up in this invalid state, so we should be fixing that instead.

Reply:
I was able to put the cursor in this position (quite by accident) with a completely different set of steps, which is why I figured it would be better to make it handle the case rather than prevent the case.

As for `moveToCaretPosition` I looked into changing it there and basically it comes down to editor schemas. There's a `moveCaretBeforeOnEnter` schema, and the code will deliberately put the cursor just _before_ any element in that schema rather than just inside if you press enter. Removing `pre` from this schema will fix the bug, but I don't know what else it will do (and it isn't a reliable fix, seeing as those schemas are technically configurable)

Reply:
For the first point, as we mentioned on a call, this is definitely not the right place to attempt to solve that (SelectionOverrides or similar is better suited, as it should be solved when it's attempting to set the selection). 

For the second point, when I looked into this, that needs to be in there for the `br_in_pre: true`. So you need to add some branching logic for that when `br_in_pre: false` is configured.

Reply:
Also adding in the br is a fair bit of a "hack" that just happens to work and gets cleaned up correctly. It's not very reliable, so I don't see it as a valid solution here sorry.